pkg/virt-controller/watch/migration.go

```diff
@@ -687,6 +687,67 @@ func (c *MigrationController) expandPDB(pdb *policyv1.PodDisruptionBudget, vmi *
 	return nil
 }
 
+func (c *MigrationController) handleMigrationBackoff(key string, vmi *virtv1.VirtualMachineInstance, migration *virtv1.VirtualMachineInstanceMigration) error {
+	if migration.Status.Phase != virtv1.MigrationPending {
+		return nil
+	}
+
+	if _, exists := migration.Annotations[virtv1.FuncTestForceIgnoreMigrationBackoffAnnotation]; exists {
+		return nil
+	}
+
+	migrations, err := c.listMigrationsMatchingVMI(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return err
+	}
+	if len(migrations) < 2 {
+		return nil
+	}
+
+	// Newest first
+	sort.Sort(sort.Reverse(vmimCollection(migrations)))
+	if migrations[0].UID != migration.UID {
+		return nil
+	}
+
+	backoff := time.Second * 0
+	for _, m := range migrations[1:] {
+		if m.Status.Phase == virtv1.MigrationSucceeded {
+			break
+		}
+
+		if m.Status.Phase == virtv1.MigrationFailed {
+			if backoff == 0 {
+				backoff = time.Second * 20
+			} else {
+				backoff = backoff * 2
+			}
+		}
+	}
+	if backoff == 0 {
+		return nil
+	}
+
+	getFailedTS := func(migration *virtv1.VirtualMachineInstanceMigration) metav1.Time {
+		for _, ts := range migrations[1].Status.PhaseTransitionTimestamps {
+			if ts.Phase == virtv1.MigrationFailed {
+				return ts.PhaseTransitionTimestamp
+			}
+		}
+		return metav1.Time{}
+	}
+
+	outOffBackoffTS := getFailedTS(migrations[1]).Add(backoff)
```

Comment:
that wouldn't be correct though, the backoff period should start from the last failed migration IMO.

Example: migration 1 failed at time T, if I create migration 2 at time T+1minute I'd like it to run immediately and not wait for the backoff time as it has already passed since T, otherwise migration 2 will run at T+1minute+backoff.

Reply:
This looks already good, I have just one last thought. Could we use the VMI`s migration state for this?

Reply:
@xpivarc Hi, you mean to use `status.migrtaionState.endTimestamp` instead of `PhaseTransitionTimestamp`?

Reply:
Yes, imho it should be a better source of truth(We would need to check if indeed we always set it, in case of timeout, cancel,...). For example, we can have 3 failed migrations in a row and the 4th one was canceled(meaning deleted). In this case, we would not do backoff but I believe VMI would have this information...

Reply:
@acardace Can we address the case of a cancelled migration?

Reply:
@enp0s3 done, thanks for pointing it out, I did not understand what you meant at first!

Reply:
@enp0s3 Was this addressed?